### PR TITLE
add ccc firehose

### PR DESCRIPTION
**Pull request by 876pol**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

**Comment by 876pol** [comment]
I'm not sure what difficulty I should list this problem as. I don't know if it should be easy or normal.

**Comment by envyaims** [comment]
i think @SansPapyrus683 said that we don't need more binary search problems. idk though

File: solutions/ccc-Firehose.mdx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+		for (int j = 1; j < num_houses; j++) {
+			// Address of house at index j.
+			int end = houses[(i + j) % num_houses];
+
+			/*
+			If the distance between the start and end houses is greater than two
+			times the length of the hose, a single fire hydrant will not be
+			enough to cover both houses.
+			*/
+			int dist = (end - start + 1000000) % 1000000;
```
Comment (envyaims) [inline solutions/ccc-Firehose.mdx:45]: make 1000000 a constant

File: solutions/ccc-Firehose.mdx [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+<LanguageSection>
+
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+int num_houses, num_hydrants, houses[1000];
```
Comment (envyaims) [inline solutions/ccc-Firehose.mdx:26]: dont use static arrays

**Review by envyaims — CHANGES_REQUESTED** [review]

File: solutions/ccc-Firehose.mdx (hunk 2 above)
Comment (envyaims) [reply solutions/ccc-Firehose.mdx:26]: nvm its fine

File: solutions/ccc-Firehose.mdx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	cin >> num_houses;
+	for (int i = 0; i < num_houses; i++) {
+		cin >> houses[i];
+	}
+	cin >> num_hydrants;
+
+	sort(houses, houses + num_houses);
+
+	// Binary search for the lowest hose length
+	int l = 0, r = STREET_SIZE, ans = 0;
```
Comment (SansPapyrus683) [inline solutions/ccc-Firehose.mdx:73]: declarations

File: solutions/ccc-Firehose.mdx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	while (l <= r) {
+		int m = (l + r) / 2;
+		if (possible(m)) {
+			r = m - 1;
+			ans = m;
+		} else {
+			l = m + 1;
+		}
+	}
+	cout << ans << "\n";
```
Comment (SansPapyrus683) [inline solutions/ccc-Firehose.mdx:83]:
```suggestion

	cout << ans << endl;
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Comment by SansPapyrus683** [comment]
> i think @SansPapyrus683 said that we don't need more binary search problems. idk though

oooh yeah...
i thought this was already in the module for some reason
i'll merge this once it's done, but please don't do any more

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/ccc-Firehose.mdx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+We use a greedy algorithm to find $\texttt{possible}(i)$. Firstly, we want to sort all houses by their address. For each $j$ from $1...n$, let $h_j$ be the address of the $j$th house. Place a fire hydrant at address $h_j + i$, and this hydrant will cover all houses in the range $[h_j, h_j + 2]$. We can repeat this process for the first house to the right whose address does not fall into this range until every house can be connected with at least one first hydrant. Note that due to the circular nature of the street, we must do all calculations$\mod 1000000$. If the number of hydrants needed is greater than the number of fire hydrants allowed, $\texttt{possible}(i)$ is true; otherwise, it is false.
+
+## Implementation
+
+**Time Complexity**: $\mathcal O(h^2\times log(1000000))$
```
Comment (envyaims) [inline solutions/ccc-Firehose.mdx:16]:
```suggestion
**Time Complexity**: $\mathcal O(h^2\times log(\texttt{max}(H_i)))$
```

**Review by envyaims — APPROVED** [review]
ok

File: solutions/ccc-Firehose.mdx [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+## Implementation
+
+**Time Complexity**: $\mathcal O(h^2\times log(\texttt{max}(H_i)))$
+
+<LanguageSection>
+
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+
```
Comment (Aadit-Ambadkar) [inline solutions/ccc-Firehose.mdx:23]:
```suggestion
```
Remove unnecessary spaces.

File: solutions/ccc-Firehose.mdx [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+			 * times the length of the hose, a single fire hydrant will not be
+			 * enough to cover both houses.
+			 */
+			int dist = (end - start + STREET_SIZE) % STREET_SIZE;
+			if (dist > 2 * length) {
+				start = end;
+				needed++;
+			}
+		}
+		// Increment needed as we need another hydrant to reach the last houses.
```
Comment (Aadit-Ambadkar) [inline solutions/ccc-Firehose.mdx:56]:
```suggestion

		// Increment needed as we need another hydrant to reach the last houses.
```
Add necessary spaces

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

**Review by Aadit-Ambadkar — APPROVED** [review]
Looks good!

**Comment by 876pol** [comment]
can this maybe be merged?

**Pull request merged**